primitives/api/proc-macro/src/runtime_metadata.rs

```diff
@@ -0,0 +1,253 @@
+		let signature = &method.sig;
+		for input in &signature.inputs {
+			// Exclude `self` from metadata collection.
+			let syn::FnArg::Typed(typed) = input else {
+				continue
+			};
+
+			let pat = &typed.pat;
+			let name = format!("{}", quote!(#pat));
+			let ty = &typed.ty;
+			collect_where_bounds(ty).map(|ty_elem| where_clause.push(ty_elem));
+
+			inputs.push(quote!(
+				#crate_::metadata_ir::RuntimeApiMethodParamMetadataIR {
+					name: #name,
+					ty: #crate_::scale_info::meta_type::<#ty>(),
+				}
+			));
+		}
+
+		let output = match &signature.output {
+			syn::ReturnType::Default => quote!(#crate_::scale_info::meta_type::<()>()),
+			syn::ReturnType::Type(_, ty) => {
+				collect_where_bounds(ty).map(|ty_elem| where_clause.push(ty_elem));
+				quote!(#crate_::scale_info::meta_type::<#ty>())
+			},
+		};
+
+		// String method name including quotes for constructing `v15::RuntimeApiMethodMetadata`.
+		let method_name = format!("{}", signature.ident);
+		let docs = collect_docs(&method.attrs, &crate_);
+
+		// Include the method metadata only if its `cfg` features are enabled.
+		let attrs = filter_cfg_attributes(&method.attrs);
+		methods.push(quote!(
+			#( #attrs )*
+			#crate_::metadata_ir::RuntimeApiMethodMetadataIR {
+				name: #method_name,
+				inputs: #crate_::vec![ #( #inputs, )* ],
+				output: #output,
+				docs: #docs,
+			}
+		));
+	}
+
+	let trait_name_ident = &decl.ident;
+	let trait_name = format!("{}", trait_name_ident);
+	let docs = collect_docs(&decl.attrs, &crate_);
+	let attrs = filter_cfg_attributes(&decl.attrs);
+	// The trait generics where already extended with `Block: BlockT`.
+	let mut generics = decl.generics.clone();
+	for generic_param in generics.params.iter_mut() {
+		let syn::GenericParam::Type(ty) = generic_param else {
+			continue
+		};
+
+		// `scale_info::meta_type` requires `T: ?Sized + TypeInfo + 'static` bounds.
+		ty.bounds.push(parse_quote!(#crate_::scale_info::TypeInfo));
+		ty.bounds.push(parse_quote!('static));
+		// Default type parameters are not allowed in functions.
+		ty.eq_token = None;
+		ty.default = None;
+	}
+
+	let where_clause: Vec<_> = where_clause
+		.iter()
+		.map(|ty| quote!(#ty: #crate_::scale_info::TypeInfo + 'static))
+		.collect();
+
+	quote!(
+		#( #attrs )*
+		#[inline(always)]
+		pub fn runtime_metadata #generics () -> #crate_::metadata_ir::RuntimeApiMetadataIR
+		where #( #where_clause, )*
+		{
+			#crate_::metadata_ir::RuntimeApiMetadataIR {
+				name: #trait_name,
+				methods: #crate_::vec![ #( #methods, )* ],
+				docs: #docs,
+			}
+		}
+	)
+}
+
+/// Implement the `runtime_metadata` function on the runtime that
+/// generates the metadata for the given traits.
+///
+/// The metadata of each trait is extracted from the generic function
+/// exposed by `generate_decl_runtime_metadata`.
+pub fn generate_impl_runtime_metadata(
+	impls: &[ItemImpl],
+	crate_: &TokenStream,
+) -> Result<TokenStream> {
+	if impls.is_empty() {
+		return Ok(quote!())
+	}
```

Comment:
You are missing the creating and implementation of `InternalImplRuntimeApis` which would then lead to a compile error.

Reply:
Yeah that was my lack of knowledge around `InternalImplRuntimeApis` :D